relay-event-normalization/src/normalize/span/description/sql/parser.rs

```diff
@@ -0,0 +1,274 @@
+use std::ops::ControlFlow;
+
+use itertools::Itertools;
+use sqlparser::ast::{
+    Expr, Ident, Query, SelectItem, SetExpr, Statement, TableFactor, UnaryOperator, Value,
+    VisitMut, VisitorMut,
+};
+use sqlparser::dialect::{Dialect, GenericDialect};
+
+/// Derive the SQL dialect from `span.system` and try to parse the query into an AST.
+pub fn parse_query(
+    db_system: Option<&str>,
+    query: &str,
+) -> Result<Vec<Statement>, sqlparser::parser::ParserError> {
+    // See https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/semantic_conventions/database.md#notes-and-well-known-identifiers-for-dbsystem
+    //     https://docs.rs/sqlparser/latest/sqlparser/dialect/fn.dialect_from_str.html
+    let dialect = db_system
+        .and_then(sqlparser::dialect::dialect_from_str)
+        .unwrap_or_else(|| Box::new(GenericDialect {}));
+    let dialect = DialectWithParameters(dialect);
+
+    sqlparser::parser::Parser::parse_sql(&dialect, query)
+}
+
+/// Tries to parse a series of SQL queries into an AST and normalize it.
+pub fn normalize_parsed_queries(db_system: Option<&str>, string: &str) -> Result<String, ()> {
+    let mut parsed = parse_query(db_system, string).map_err(|_| ())?;
+    parsed.visit(&mut NormalizeVisitor);
+
+    Ok(parsed
+        .iter()
+        .map(|statement| statement.to_string())
+        .join("; "))
+}
+
+/// A visitor that normalizes the SQL AST in-place.
+///
+/// Used for span description normalization.
+pub struct NormalizeVisitor;
+
+impl NormalizeVisitor {
+    /// Placeholder for string and numerical literals.
+    fn placeholder() -> Value {
+        Value::Number("%s".into(), false)
+    }
+
+    /// Placeholder for a list of items that has been scrubbed.
+    fn ellipsis() -> Ident {
+        Ident::new("..")
+    }
+
+    /// Check if a selected item can be erased into `..`. Currently we only collapse simple
+    /// columns (`col1` or `col1 AS foo`).
+    fn is_collapsible(item: &SelectItem) -> bool {
+        match item {
+            SelectItem::UnnamedExpr(expr) => {
+                matches!(expr, Expr::Identifier(_) | Expr::CompoundIdentifier(_))
+            }
+            SelectItem::ExprWithAlias { expr, .. } => {
+                matches!(expr, Expr::Identifier(_) | Expr::CompoundIdentifier(_))
+            }
+            _ => false,
+        }
+    }
+
+    /// Helper function to collapse a sequence of selected items into `..`.
+    fn collapse_items(collapse: &mut Vec<SelectItem>, output: &mut Vec<SelectItem>) {
+        match collapse.len() {
+            0 => {}
+            1 => {
+                output.append(collapse);
+            }
+            _ => {
+                output.push(SelectItem::UnnamedExpr(Expr::Identifier(Self::ellipsis())));
+            }
+        }
+    }
+
+    /// Normalizes `SELECT ...` queries.
+    fn transform_query(query: &mut Query) {
```

Comment:
nit: I'd split this function into different ones for each keyword (e.g. select, from).